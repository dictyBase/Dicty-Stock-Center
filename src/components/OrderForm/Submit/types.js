export type Props = {
  /** Formik value to indicate if form is being submitted */
  isSubmitting: boolean,
  /** Current order form page number */
  pageNum: number,
  /** Function to set the page number */
  setPageNum: Function,
}
