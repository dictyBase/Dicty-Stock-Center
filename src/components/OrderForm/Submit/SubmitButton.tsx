import React from "react"
import { useFormikContext } from "formik"
import Button from "@material-ui/core/Button"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { useMutation, useLazyQuery } from "@apollo/react-hooks"
import { useHistory } from "react-router-dom"
import { useCartStore } from "components/ShoppingCart/CartStore"
import useCartItems from "hooks/useCartItems"
import { CREATE_ORDER, CREATE_USER, UPDATE_USER } from "graphql/mutations"
import { GET_USER_BY_EMAIL } from "graphql/queries"
import useStyles from "../formStyles"
import { FormikValues } from "../utils/initialValues"
import { CartItem } from "../types"

const getIDs = (items: Array<CartItem>) =>
  items.map((item: CartItem) => item.id)

/**
 * getUserVariables generates a variables object that is passed with
 * create or update user mutations.
 */
const getUserVariables = (values: FormikValues, id?: string) => {
  const variablesObj: any = {
    variables: {
      input: {
        first_name: values.firstName,
        last_name: values.lastName,
        email: values.email,
        organization: values.organization,
        group_name: values.lab,
        first_address: values.address1,
        second_address: values.address2,
        city: values.city,
        state: values.state,
        zipcode: values.zip,
        country: values.country,
        phone: values.phone,
        is_active: true,
      },
    },
  }
  // if there's an ID, that means we need to send an update mutation;
  // add ID to variables and remove email (since this cannot be updated)
  if (id) {
    variablesObj.variables.id = id
    delete variablesObj.variables.input.email
  }
  return variablesObj
}

/**
 * getOrderVariables generates a variables object that is sent with the
 * create order mutation.
 */
const getOrderVariables = (
  values: FormikValues,
  addedItems: Array<CartItem>,
) => ({
  variables: {
    input: {
      courier: values.shippingAccount,
      courier_account: values.shippingAccountNumber,
      comments: values.comments,
      payment: values.paymentMethod,
      purchase_order_num: values.purchaseOrderNum,
      status: "IN_PREPARATION",
      consumer: values.email,
      payer: values.payerEmail,
      purchaser: values.email,
      items: getIDs(addedItems),
    },
  },
})

/**
 * SubmitButton is the button used to submit the order. It
 * appears on the last page of the order form, and it contains the
 * necessary logic for GraphQL queries and mutations.
 */

const SubmitButton = ({ setSubmitError }: { setSubmitError: Function }) => {
  const { isSubmitting, values, submitForm } = useFormikContext<any>()
  const [{ addedItems }] = useCartStore()
  const { emptyCart } = useCartItems(addedItems)
  const history = useHistory()
  const [createOrder] = useMutation(CREATE_ORDER, {
    onCompleted: () => {
      setSubmitError(false)
      submitForm()
      history.push("/order/submitted")
      emptyCart()
    },
    onError: () => {
      setSubmitError(true)
    },
  })
  const [createUser] = useMutation(CREATE_USER, {
    onCompleted: () => {
      setSubmitError(false)
      createOrder(getOrderVariables(values, addedItems))
    },
    onError: () => {
      setSubmitError(true)
    },
  })
  const [updateUser] = useMutation(UPDATE_USER, {
    onCompleted: () => {
      setSubmitError(false)
      createOrder(getOrderVariables(values, addedItems))
    },
    onError: () => {
      setSubmitError(true)
    },
  })
  const [getUserByEmail] = useLazyQuery(GET_USER_BY_EMAIL, {
    onCompleted: (data) => {
      updateUser(getUserVariables(values, data.userByEmail.id))
      setSubmitError(false)
    },
    onError: (error: any) => {
      if (
        error.graphQLErrors.length > 0 &&
        error.graphQLErrors[0].extensions.code === "NotFound"
      ) {
        setSubmitError(false)
        createUser(getUserVariables(values))
        return
      }
      setSubmitError(true)
    },
  })
  const classes = useStyles()

  const handleSubmit = () => {
    getUserByEmail({
      variables: { email: values.email },
    })
  }

  return (
    <Button
      type="submit"
      size="large"
      className={classes.submitBtn}
      disabled={isSubmitting}
      onClick={handleSubmit}>
      Submit Order &nbsp;
      <FontAwesomeIcon icon="check-circle" />
    </Button>
  )
}

export default SubmitButton
