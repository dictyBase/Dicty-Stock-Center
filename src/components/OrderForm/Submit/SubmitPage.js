// @flow
import React from "react"
import { withStyles } from "@material-ui/core/styles"
import Grid from "@material-ui/core/Grid"
import Button from "@material-ui/core/Button"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import ShoppingCartItems from "components/ShoppingCart/ShoppingCartItems"
import styles from "../formStyles"

type Props = {
  /** Material-UI styling */
  classes: Object,
  /** Formik value to indicate if form is being submitted */
  isSubmitting: boolean,
  /** Current order form page number */
  pageNum: number,
  /** Function to set the page number */
  setPageNum: Function,
}

/**
 * SubmitPage is the final page the user sees before submitting the order.
 */

export const SubmitPage = (props: Props) => {
  const { classes, pageNum, setPageNum } = props

  return (
    <Grid container spacing={16} className={classes.innerForm}>
      <Grid item xs={12}>
        <div className={classes.header}>Please review your order</div>
      </Grid>
      <Grid item xs={12}>
        <div className={classes.submitPage}>
          <h1>
            <FontAwesomeIcon icon="truck" /> Items
          </h1>
          <Grid container justify="center">
            <ShoppingCartItems />
          </Grid>
        </div>
        <br />
        <br />
        <br />
        <Grid container justify="center" spacing={16}>
          <Grid item xs={2} />
          <Grid item xs={4}>
            <Button
              color="primary"
              size="large"
              className={classes.previousBtn}
              onClick={() => setPageNum(pageNum - 1)}>
              <FontAwesomeIcon icon="arrow-circle-left" />
              &nbsp; Back
            </Button>
          </Grid>
          <Grid item xs={4}>
            <Button
              type="submit"
              size="large"
              className={classes.submitBtn}
              disabled={props.isSubmitting}>
              Submit Order &nbsp;
              <FontAwesomeIcon icon="check-circle" />
            </Button>
          </Grid>
          <Grid item xs={2} />
        </Grid>
      </Grid>
    </Grid>
  )
}

export default withStyles(styles)(SubmitPage)
