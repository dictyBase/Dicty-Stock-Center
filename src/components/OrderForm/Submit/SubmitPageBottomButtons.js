// @flow
import React from "react"
import Grid from "@material-ui/core/Grid"
import Button from "@material-ui/core/Button"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import BackButton from "../BackButton"
import useStyles from "../formStyles"
import { Props } from "./types"

/**
 * SubmitPageBottomButtons is the display for the buttons at the bottom of
 * the final submit page.
 */

const SubmitPageBottomButtons = ({
  pageNum,
  setPageNum,
  isSubmitting,
}: Props) => {
  const classes = useStyles()

  return (
    <Grid container justify="center" spacing={2}>
      <Grid item xs={2} />
      <Grid item xs={4}>
        <BackButton pageNum={pageNum} setPageNum={setPageNum} />
      </Grid>
      <Grid item xs={4}>
        <Button
          type="submit"
          size="large"
          className={classes.submitBtn}
          disabled={isSubmitting}>
          Submit Order &nbsp;
          <FontAwesomeIcon icon="check-circle" />
        </Button>
      </Grid>
      <Grid item xs={2} />
    </Grid>
  )
}

export default SubmitPageBottomButtons
