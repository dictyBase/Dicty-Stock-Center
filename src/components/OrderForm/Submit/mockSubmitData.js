export const data = [
  {
    id: "DBS123",
    name: "test1",
  },
  {
    id: "DBS456",
    name: "test2",
  },
]
