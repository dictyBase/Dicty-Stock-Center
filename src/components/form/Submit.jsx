// @flow
import React, { Component } from "react"
import { connect } from "react-redux"
import EditPanel from "./EditPanel"
import Items from "components/Items"
import { editShipping } from "actions/order/shipping"
import { editPayment } from "actions/order/payment"
import { submitOrder } from "actions/order/submit"
import { Panel, PanelBody } from "dicty-components-panel"
import { Flex, Box } from "rebass"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { DictyHeader, PrimaryLargeButton } from "styles"

type Props = {
  order: Object,
  cart: Object,
  editShipping: Function,
  editPayment: Function,
  submitOrder: Function,
  addedItems: Array<Object>,
  consumer: Object,
  shipping: Object,
  payer: Object,
  payment: Object,
}

export class Submit extends Component<Props> {
  render() {
    const {
      addedItems,
      consumer,
      shipping,
      payer,
      payment,
      editShipping,
      editPayment,
      submitOrder,
    } = this.props
    return (
      <Flex wrap justify="center">
        <Box w={["85%", "60%"]}>
          <DictyHeader>
            <h2>Review Your Order</h2>
            <p>
              {" "}
              Please review your order and click "Complete Your Order" button to
              place your order
            </p>
          </DictyHeader>
        </Box>
        <Box w={["85%", 3 / 4]}>
          <Flex wrap justify="center">
            <Box w={[1, 1, 1, "45%"]} mr={1}>
              <Panel collapse>
                <PanelBody>
                  <EditPanel
                    user={consumer}
                    edit={editShipping}
                    title={" Ship to:"}
                    icon={"truck"}
                    shipping={shipping}
                  />
                </PanelBody>
              </Panel>
            </Box>
            <Box w={[1, 1, 1, "45%"]} mr={1}>
              <Panel collapse>
                <PanelBody>
                  <EditPanel
                    user={payer}
                    edit={editPayment}
                    title={" Payer:"}
                    icon={"credit-card-alt"}
                    payment={payment}
                  />
                </PanelBody>
              </Panel>
            </Box>
          </Flex>
          <hr />
          <Flex wrap justify="center">
            <Box w={1} mr={1}>
              <Panel collapse>
                <PanelBody>
                  <h4>
                    <FontAwesomeIcon icon="shopping-cart" /> Items
                  </h4>
                  <Items items={addedItems} />
                </PanelBody>
              </Panel>
            </Box>
            <Box w={1} mb={1} mr={1}>
              <hr />
              <PrimaryLargeButton type="button" onClick={submitOrder}>
                Complete Your Order <FontAwesomeIcon icon="check-circle" />
              </PrimaryLargeButton>
            </Box>
          </Flex>
        </Box>
      </Flex>
    )
  }
}

const mapStateToProps = state => ({
  addedItems: state.cart.addedItems,
  consumer: state.order.consumer,
  shipping: state.order.shipping,
  payer: state.order.payer,
  payment: state.order.payment,
})

const mapDispatchToProps = dispatch => ({
  editShipping: () => {
    dispatch(editShipping())
  },
  editPayment: () => {
    dispatch(editPayment())
  },
  submitOrder: () => {
    dispatch(submitOrder())
  },
})

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Submit)
