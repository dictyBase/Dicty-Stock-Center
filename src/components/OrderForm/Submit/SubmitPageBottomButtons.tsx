import React from "react"
import Grid from "@material-ui/core/Grid"
import BackButton from "../BackButton"
import SubmitButton from "./SubmitButton"
import { PageProps } from "../types"

/**
 * SubmitPageBottomButtons is the display for the buttons at the bottom of
 * the final submit page.
 */

const SubmitPageBottomButtons = ({
  pageNum,
  setPageNum,
  setSubmitError,
}: PageProps) => (
  <Grid container justify="center" spacing={2}>
    <Grid item xs={2} />
    <Grid item xs={4}>
      <BackButton pageNum={pageNum} setPageNum={setPageNum} />
    </Grid>
    <Grid item xs={4}>
      <SubmitButton setSubmitError={setSubmitError} />
    </Grid>
    <Grid item xs={2} />
  </Grid>
)

export default SubmitPageBottomButtons
