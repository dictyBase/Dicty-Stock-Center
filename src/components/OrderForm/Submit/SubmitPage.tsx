import React from "react"
import Grid from "@material-ui/core/Grid"
import ShoppingCartItemList from "components/ShoppingCart/ShoppingCartItemList"
import SubmitPageBottomButtons from "./SubmitPageBottomButtons"
import useStyles from "../formStyles"
import { Props } from "./types"

/**
 * SubmitPage is the final page the user sees before submitting the order.
 */

export const SubmitPage = ({ pageNum, setPageNum }: Props) => {
  const classes = useStyles()

  return (
    <Grid container spacing={2} className={classes.innerForm}>
      <Grid item xs={12}>
        <div className={classes.submitPage}>
          <Grid container justify="center">
            <ShoppingCartItemList />
          </Grid>
        </div>
        <br />
        <br />
        <br />
        <SubmitPageBottomButtons pageNum={pageNum} setPageNum={setPageNum} />
      </Grid>
    </Grid>
  )
}

export default SubmitPage
