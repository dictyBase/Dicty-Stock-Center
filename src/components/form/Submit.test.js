import React from "react"
import renderer from "react-test-renderer"
import "jest-styled-components"
import { Submit } from "./Submit"

test("matching a snapshot of Submit", () => {
  const editShipping = () => {}
  const editPayment = () => {}
  const addedItems = ["1", "2"]
  const consumer = {
    firstName: "Jane"
  }
  const payer = {
    firstName: "John"
  }
  const component = renderer.create(
    <Submit
      editShipping={editShipping}
      editPayment={editPayment}
      addedItems={addedItems}
      consumer={consumer}
      payer={payer}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
