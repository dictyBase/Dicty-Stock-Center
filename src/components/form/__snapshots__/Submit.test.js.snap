// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matching a snapshot of Submit 1`] = `
<div
  className="sc-bxivhb bUZGRr sc-htpNat eDtRyC sc-bdVaJa iHZvIS"
  is={null}
>
  <div
    className="sc-ifAKCX cFlEyZ sc-bdVaJa dEKHPo"
    is={null}
  >
    <div
      className="sc-TFwJa daqYjB"
    >
      <h2>
        Review Your Order
      </h2>
      <p>
         
        Please review your order and click "Complete Your Order" button to place your order
      </p>
    </div>
  </div>
  <div
    className="sc-ifAKCX cFlEyZ sc-bdVaJa foxScs"
    is={null}
  >
    <div
      className="sc-bxivhb bUZGRr sc-htpNat eDtRyC sc-bdVaJa iHZvIS"
      is={null}
    >
      <div
        className="sc-ifAKCX cFlEyZ sc-bdVaJa vDLMO"
        is={null}
      >
        <div
          className="gsdTnU"
          style={
            Object {
              "marginBottom": 0,
            }
          }
        >
          <div
            className="panel-body cgBkWQ"
            open={true}
            style={Object {}}
          >
            <div
              className="DRXhs"
            />
            <div>
              <h4>
                <span
                  aria-hidden={true}
                  className="fa fa-truck"
                />
                 Ship to:
              </h4>
              <div>
                Jane
                 
              </div>
              <div />
              <div />
              <div>
                 
                 
              </div>
              <div />
              <br />
              <div>
                <a
                  className="sc-giadOv kWtcoa"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden={true}
                    className="fa fa-pencil-square-o"
                  />
                   Edit shipping info
                </a>
              </div>
            </div>
            <div
              className="DRXhs"
            />
          </div>
        </div>
      </div>
      <div
        className="sc-ifAKCX cFlEyZ sc-bdVaJa vDLMO"
        is={null}
      >
        <div
          className="gsdTnU"
          style={
            Object {
              "marginBottom": 0,
            }
          }
        >
          <div
            className="panel-body cgBkWQ"
            open={true}
            style={Object {}}
          >
            <div
              className="DRXhs"
            />
            <div>
              <h4>
                <span
                  aria-hidden={true}
                  className="fa fa-credit-card-alt"
                />
                 Payer:
              </h4>
              <div>
                John
                 
              </div>
              <div />
              <div />
              <div>
                 
                 
              </div>
              <div />
              <br />
              <div>
                <a
                  className="sc-giadOv kWtcoa"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden={true}
                    className="fa fa-pencil-square-o"
                  />
                   Edit shipping info
                </a>
              </div>
            </div>
            <div
              className="DRXhs"
            />
          </div>
        </div>
      </div>
    </div>
    <hr />
    <div
      className="sc-bxivhb bUZGRr sc-htpNat eDtRyC sc-bdVaJa iHZvIS"
      is={null}
    >
      <div
        className="sc-ifAKCX cFlEyZ sc-bdVaJa fiWxJc"
        is={null}
      >
        <div
          className="gsdTnU"
          style={
            Object {
              "marginBottom": 0,
            }
          }
        >
          <div
            className="panel-body cgBkWQ"
            open={true}
            style={Object {}}
          >
            <div
              className="DRXhs"
            />
            <h4>
              <span
                aria-hidden={true}
                className="fa fa-shopping-cart"
              />
               Items
            </h4>
            <div>
              <div
                className="sc-bxivhb bUZGRr sc-htpNat jtWIOA sc-bdVaJa iHZvIS"
                is={null}
              >
                <div
                  className="sc-ifAKCX cFlEyZ sc-bdVaJa goNVXm"
                  is={null}
                >
                  <h5>
                    ID
                  </h5>
                </div>
                <div
                  className="sc-ifAKCX cFlEyZ sc-bdVaJa goNVXm"
                  is={null}
                >
                  <h5>
                    Strain/Plasmid Name
                  </h5>
                </div>
              </div>
              <div
                className="sc-bxivhb bUZGRr sc-htpNat jtWIOA sc-bdVaJa iHZvIS"
                is={null}
              >
                <div
                  className="sc-ifAKCX cFlEyZ sc-bdVaJa goNVXm"
                  is={null}
                />
                <div
                  className="sc-ifAKCX cFlEyZ sc-bdVaJa goNVXm"
                  is={null}
                />
              </div>
              <div
                className="sc-bxivhb bUZGRr sc-htpNat jtWIOA sc-bdVaJa iHZvIS"
                is={null}
              >
                <div
                  className="sc-ifAKCX cFlEyZ sc-bdVaJa goNVXm"
                  is={null}
                />
                <div
                  className="sc-ifAKCX cFlEyZ sc-bdVaJa goNVXm"
                  is={null}
                />
              </div>
            </div>
            <div
              className="DRXhs"
            />
          </div>
        </div>
      </div>
      <div
        className="sc-ifAKCX cFlEyZ sc-bdVaJa kIMTZA"
        is={null}
      >
        <hr />
        <button
          className="large block sc-kaNhvL jpxKph"
          onClick={undefined}
          type="button"
        >
          Complete Your Order 
          <span
            aria-hidden={true}
            className="fa fa-check-circle"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
